File: python/p14.py
#!/usr/bin/python

#The following iterative sequence is defined for the set of positive integers:
#
#n  n/2 (n is even)
#n  3n + 1 (n is odd)
#
#Using the rule above and starting with 13, we generate the following sequence:
#
#13  40  20  10  5  16  8  4  2  1
#It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.
#
#Which starting number, under one million, produces the longest chain?
#
#NOTE: Once the chain starts the terms are allowed to go above one million.



#naive
startNum=13

def chain(num):
	chainlength = 1
	while num > 1:
		if num%2 == 0:
			num/=2
		else:
			num=num*3+1
		chainlength+=1
	return chainlength 
	
maxLength = (1,1)
for i in range(500000,1000000):
	res = (chain(i),i)
	maxLength = max(res,maxLength)

print maxLength

File: c++/p14.cpp
#include "p14.hpp"
#include <vector>
#include <stack>

//The following iterative sequence is defined for the set of positive integers:
//
//n ->  n/2 (n is even)
//n ->  3n + 1 (n is odd)
//
//Which starting number, under one million, produces the longest chain?
//
//
static std::vector<uint> answerMap;

uint collatz(uint start)
{
  std::stack<uint> steps;
  uint curStep = start;
  steps.push(curStep);

  while ( curStep != 1 )
  {
    if ( answerMap.size() > curStep && answerMap[curStep] != 0 )
    {
      steps.push(curStep);
      break;
    }
    if ( curStep % 2 == 0 )
      curStep /= 2;
    else
      curStep = 3 * curStep + 1;
    steps.push(curStep);
  }

  uint answerSteps = answerMap[steps.top()];
  steps.pop();

  while ( steps.empty() == false )
  {
    if ( steps.top() < answerMap.size() )
      answerMap[steps.top()] = answerSteps;
    steps.pop();
    ++answerSteps;
  }

  return answerMap[start];
}

uint p14::execute() 
{
  answerMap.resize(1000000,0);
  answerMap[0] = answerMap[1] = 1;
  
  uint answer = 1, maxStepsFound = 1;

  for ( uint i = 1; i < 1000000; ++i )
  {
    uint col = collatz(i);
    if ( maxStepsFound < col )
    {
      answer = i;
      maxStepsFound = col;
    }
  }
  return answer;
}
